chainer/functions/pooling/upsampling_2d.py

```diff
@@ -0,0 +1,219 @@
+        self.indexes = indexes
+        self.outh, self.outw = (None, None) if outsize is None else outsize
+
+    def check_type_forward(self, in_types):
+        n_in = in_types.size()
+        type_check.expect(n_in == 1)
+        x_type = in_types[0]
+
+        type_check.expect(
+            x_type.dtype.kind == 'f',
+            x_type.ndim == 4,
+            x_type.shape == self.indexes.shape,
+        )
+
+        if self.outh is not None:
+            expected_h = conv.get_conv_outsize(
+                self.outh, self.kh, self.sy, self.ph, cover_all=self.cover_all)
+            type_check.expect(x_type.shape[2] == expected_h)
+        if self.outw is not None:
+            expected_w = conv.get_conv_outsize(
+                self.outw, self.kw, self.sx, self.pw, cover_all=self.cover_all)
+            type_check.expect(x_type.shape[3] == expected_w)
+
+    def forward_cpu(self, x):
+        n, c, h, w = x[0].shape
+        if self.outh is None:
+            self.outh = conv.get_deconv_outsize(
+                h, self.kh, self.sy, self.ph, cover_all=self.cover_all)
+        if self.outw is None:
+            self.outw = conv.get_deconv_outsize(
+                w, self.kw, self.sx, self.pw, cover_all=self.cover_all)
+
+        up_y = numpy.zeros((n, c, self.outh, self.outw), dtype=numpy.float32)
+        up_y = conv.im2col_cpu(
+            up_y, self.kh, self.kw, self.sy, self.sx, self.ph, self.pw,
+            cover_all=self.cover_all)
+        for n in six.moves.range(up_y.shape[0]):
+            for c in six.moves.range(up_y.shape[1]):
+                for ky in six.moves.range(up_y.shape[4]):
+                    for kx in six.moves.range(up_y.shape[5]):
+                        sy = self.indexes[n, c, ky, kx] // up_y.shape[3]
+                        sx = self.indexes[n, c, ky, kx] % up_y.shape[3]
+                        up_y[n, c, sy, sx, ky, kx] = x[0][n, c, ky, kx]
+        up_y = conv.col2im_cpu(up_y, self.sy, self.sx, self.ph,
+                               self.pw, self.outh, self.outw)
+        return up_y,
+
+    def forward_gpu(self, x):
+        xp = cuda.cupy
+        n, c, h, w = x[0].shape
+        if self.outh is None:
+            self.outh = conv.get_deconv_outsize(
+                h, self.kh, self.sy, self.ph, cover_all=self.cover_all)
+        if self.outw is None:
+            self.outw = conv.get_deconv_outsize(
+                w, self.kw, self.sx, self.pw, cover_all=self.cover_all)
+        up_y = xp.zeros((n, c, self.outh, self.outw), dtype=numpy.float32)
+        up_y = conv.im2col_gpu(
+            up_y, self.kh, self.kw, self.sy, self.sx, self.ph, self.pw,
+            cover_all=self.cover_all)
+        up_y = up_y.transpose(0, 1, 4, 5, 2, 3)
+        n, c, sy, sx, ky, kx = up_y.shape
+        indexes = xp.asarray(self.indexes, dtype=numpy.int32)
+        xp.ElementwiseKernel(
+            'int32 indexes, float32 x, int32 n, int32 c, int32 sy, int32 sx,'
```

Comment:
It should be ``index`` instead of ``indexes``. Because it is an element of ``indexes``.